drone_flightplan/terrain_following_waylines.py

```diff
@@ -0,0 +1,277 @@
+#!/usr/bin/python3
+"""
+Convert a terrain following drone waypoint mission to a terrain following wayline mission by removing as many waypoints as can be done without deviating beyond a certain threshold from the desired Altitude Above Ground Level.
+"""
+
+import sys, os
+import argparse
+import json
+from shapely.geometry import shape
+from shapely import distance
+from shapely.ops import transform
+from pyproj import Transformer
+
+import csv
+
+def extract_lines(plan):
+    """
+    Return the waypoints of a flight plan as individual unidirectional lines.
+    Works by inspecting the heading property of each point and creating
+    a new line each time the direction changes.
+
+    Parameter
+    --------
+    plan : list
+        The 'features' portion of a GeoJSON flight plan from the Drone Tasking
+        Manager. This consists of a list of waypoints (which are dicts)
+
+    Returns
+    --------
+    waylines : list
+        A list of lists containing waypoints (which are dicts)
+    
+    """
+    # Empty list to contain all of the lines in the flight plan, each line
+    # defined as a series of points proceeding along the same heading
+    waylines = []
+    
+    # Empty list to hold the points of an individual unidirectional line
+    currentline = []
+    lastheading = None
+
+    for point in plan:
+        currentheading = point['properties']['heading']
+        # If it's not the first line, and the direction has changed
+        if lastheading and currentheading != lastheading:
+            # Yep, changed direction. Add the current line to the list of lines
+            # and flush the current line before adding the current point.
+            waylines.append(currentline)
+            currentline = []
+        currentline.append(point)
+        lastheading = currentheading
+    return waylines
+
+#def trim(line, threshold):
+def trim(line, threshold, writer):
+    """
+    Return a wayline flight line, with the first and last point intact
+    but as many as possible of the intermediate points removed
+    consistent with not exceeding the threshold of deviation from AGL.
+
+    Parameters
+    --------
+    line : list
+        A waypoint flight line as a list of waypoints (dicts)
+
+    threshold : float
+        The allowable deviation from a consistent AGL in m
+
+    writer : csv.writer object
+        Here only passing through to the inject function
+
+    Returns
+    --------
+    new_line : list
+        A wayline as list of waypoints (dicts), hopefully indistinguishable
+        from the input line, but with fewer points while retaining the
+        AGL within the threshold
+    """
+    # Work in meters. Assumes input is in EPSG:4326 (will break if not).
```

Comment:
You could always do a rough check here to see if the coordinates are valid EPSG:4326, else throw an error

```python
coord = line[0]["geometry"]["coordinates"][0]
if coord > 180 or coord < -180:
    msg = f"Coordinates are not valid EPSG. First coord {coord}"
    print(msg)
    raise ValueError(msg)
```